index.md

```diff
@@ -0,0 +1,287 @@
+
+Here is example Python code to do this from a complex real-world project:
+
+```python
+import csv
+from pathlib import Path
+import subprocess
+
+class Recorder:
+    def __init__(self, func):
+        path = Path(inspect.getfile(func)).parent
+        self.path = path / "golden" / (func.__qualname__ + ".csv")
+    # ...
+    def write(self):
+        # ...
+        with self.path.open("w") as f:
+            writer = csv.DictWriter(f, fieldnames=headers)
+            writer.writeheader()
+            writer.writerows(data)
+        if subprocess.check_output(["git", "status", "--porcelain", self.path.absolute()]):
+            subprocess.call(f"git diff {self.path.absolute()} | tail -n 4", shell=True)
+            assert False, f"differences in file {self.path.absolute()}"
+```
+
+### DRY Expectations best practices
+
+Treat your test code like code and Don't Repeat Yourself.
+
+Your test case specification API should be a simple and well designed [DSL](https://en.wikipedia.org/wiki/Domain-specific_language), either embedded or in rare cases external (parsed), that allows you to specify only the crux of what you want to test, ideally in one line per case. Go all-in on [convention over configuration](https://en.wikipedia.org/wiki/Convention_over_configuration). Put thought into it and evolve it as needed.
+
+Writing a test case should take seconds at most. Copying a case and changing one detail should be trivial (and done often, because tests should usually come in pairs that tightly hug a boundary in the space of inputs). However, if most of your cases repeat a lot of information, there should be a way to abstract that out and, in each test case, only write out the part that varies between tests.
+
+Your output format should be very readable. Put thought into it and evolve it as needed.
+
+It's okay to have more than one output format for different types of tests (e.g. in HyperFormula we might want one output format for parser tests and one for calculation engine tests), but try to have just a few that give you enough power to be able to test everything about the project.
+
+Your output format should play well with diffs (so ideally a line per thing-that-might-change, but apply common sense and non-default diffing tools if your output is e.g. naturally [tabular](https://github.com/paulfitz/daff) or a [bitmap](https://github.com/mapbox/pixelmatch)).
```

Comment:
One relevant annoyance I've seen with Jest snapshots is renaming an existing test/snapshot changes their order in the golden file (Jest sorts by name), which makes the change hard to review.  

I imagine `git diff --color-moved` etc. may help, but the reality was we were barely looking at the snapshots anyway, so anything not "in your face" in GitLab/GitHub would just be blindly trusted :see_no_evil: 

=> IMHO "play well with diffs" is crucial for approach's success, so if you have any more detailed advice here, worth elaborating.

Reply:
The same advice from before applies - run tests and commit, rename tests, run tests and commit. I may edit this to its own section so it covers this too.